src/Kestrel.Core/Internal/Http2/HPack/IntegerDecoder.cs

```diff
@@ -17,24 +19,33 @@ public bool BeginDecode(byte b, int prefixLength)
                 Value = b;
                 return true;
             }
-            else
-            {
-                _i = b;
-                _m = 0;
-                return false;
-            }
+
+            _i = b;
+            _m = 0;
+            return false;
         }
 
         public bool Decode(byte b)
         {
-            _i = _i + (b & 127) * (1 << _m);
+            _i = _i + (b & 0x7fu) * (1u << _m);
             _m = _m + 7;
 
-            if ((b & 128) != 128)
+            if ((b & 0x80) != 0x80)
             {
-                Value = _i;
+                // Int32.MaxValue only needs a maximum of 5 bytes to represent and the last byte cannot have any value set larger than 0x7
+                if ((_m > 28 && (b & 0xf8) > 0) || _i > int.MaxValue)
+                {
+                    throw new HPackDecodingException("Integer too big");
+                }
+
+                Value = unchecked((int)_i);
                 return true;
             }
+            else if (_m > 28)
```

Comment:
I'm explicitly excluding encodings that has extra empty continuation bytes. Technically it's possible to send an infinite amount of empty continuation bytes `0x80` without actually increasing the encoded integer's value.